Add scale out functionality for SNO Cluster (#689)
redhat-performance/jetlag: docs/scale-out-sno.md

# Scale Out a Single-Node Openshift Deployment

A JetLag deployed Single-Node Openshift deployment can be scaled out via JetLag. Workers can be added using JetLag Inventory and Playbooks. This guide assumes you have an existing Single-Node OCP cluster deployed via JetLag. The worker section in the JetLag inventory file should contain records that represent the worker nodes currently joined to the running cluster.

_**Steps to Scale Out:**_
- [Add New Node Entries to Worker Inventory](#add-new-node-entries-to-worker-inventory)
- [Update scale_out.yml](#update-scale_out.yml)
- [Run mno-scale-out.yml](#run-mno-scale-out.yml)

## Add Nodes to Worker Inventory

> You can add new entries to the worker inventory section manually. Place them at the end of the list of worker entries.

Author: Good Idea...Let me look into it.

- scale_out_count: This value indicates the number of entries in the worker inventory section that will be added as new workers to the existing cluster.

Example: If the initial OCP deployment had three baremetal workers and the intended worker count was ten, current_worker_count would be 3 and scale_out_count would be 7. Scale out from three existing workers, adding seven new workers, for a total of ten worker nodes.

## Run mno-scale-out.yml
Once the new worker records are added and Worker node variables are properly populatedt to the inventory and the scale_out.yml file has the proper values. The final step is to run the mno-scale-out.yml playbook.

```console
(.ansible) [root@xxx-h01-000-r650 jetlag]# ansible-playbook -i ansible/inventory/cloud99.local ansible/mno-scale-out.yml
...
```

This playbook will:
- Generate node configuration yml
- Invoke ```oc adm node-image create``` with the node configuration, which generates a discovery ISO
- Boot the new worker nodes off of the generated discovery ISO

> - Approved generated CSRs

Reviewer: typo: "Approve"

Author: Let me update as suggested.

# Scale Out a Single-Node Openshift Deployment

A JetLag deployed Single-Node Openshift deployment can be scaled out via JetLag. Workers can be added using JetLag Inventory and Playbooks. This guide assumes you have an existing Single-Node OCP cluster deployed via JetLag. The worker section in the JetLag inventory file should contain records that represent the worker nodes currently joined to the running cluster.

_**Steps to Scale Out:**_
- [Add New Node Entries to Worker Inventory](#add-new-node-entries-to-worker-inventory)
- [Update scale_out.yml](#update-scale_out.yml)
- [Run mno-scale-out.yml](#run-mno-scale-out.yml)

## Add Nodes to Worker Inventory

> To add new node entries to the worker inventory there are two potential options.
> 
> 1. New bare metal nodes are to be added to SNO Cluster
> 
>    If more nodes were added to SNO Cluster, update worker_node_count in the ansible/vars/all.yml file and rerun the create-inventory playbook. Be sure to compare the previous inventory file to the new one to ensure that everything is the same except the new nodes added to the worker section. Make sure to populate the extra vars related to Worker nodes.
> 
> 2. Manual entry
> 
>    You can add new entries to the worker inventory section manually. Place them at the end of the list of worker entries.
> 
>    The new bare metal nodes, must be placed at the end of the worker nodes inventory. The scale out playbook is designed to use the last n nodes in the inventory.
> 
>    Populate the worker node vars([worker:vars]) same as SNO node vars ([sno:vars]) and update the role parameter as worker in [worker:vars].

Reviewer 2: ```suggestion
There are two methods for adding new worker node entries to the inventory. Choose one based on your workflow:

### Option 1: Regenerate Inventory via Ansible (Recommended for SNO Additions)

Use this method if you're adding new bare metal nodes to a **Single Node OpenShift (SNO)** cluster.

1. Update the `worker_node_count` value in `ansible/vars/all.yml` to reflect the new total number of worker nodes.
2. Rerun the `create-inventory` playbook.
3. After generation, compare the updated inventory file with the previous version to ensure only the new nodes were added to the `[worker]` section.
4. Ensure all required worker-related variables are set in `[worker:vars]`.

### Option 2: Manual Inventory Update

Use this method to manually add new nodes without rerunning the playbook.

1. Append the new worker nodes to the **end** of the `[worker]` section in the inventory.
   - ⚠️ The **scale-out playbook** uses the **last `n` nodes** in the list, so order matters.
2. In the `[worker:vars]` section:
   - Populate the required variables just as you would in `[sno:vars]`.
   - Set the `role` parameter to `worker`.
Here is an example:

    [worker]
    e29-xxx-xxxx-r640 bmc_address=**REDACTED** mac_address=**REDACTED**       lab_mac=**REDACTED** ip=**REDACTED** vendor=Dell install_disk=/dev/sda

    [worker:vars]
    role=worker
    bmc_user=**REDACTED**
    bmc_password=**REDACTED**
    lab_interface=**REDACTED**
    network_interface=**REDACTED**
    network_prefix=**REDACTED**
    gateway=**REDACTED**
    dns1=**REDACTED**

```

## Update scale_out.yml
There are two variables in ansible/vars/scale_out.yml that indicate which entries from the worker inventory section should be added to the existing cluster.

- current_worker_count: This value indicates the number of entries in the worker inventory section to skip before starting to add nodes to the existing cluster. This number should match the current number of worker nodes associated with the existing cluster.
- scale_out_count: This value indicates the number of entries in the worker inventory section that will be added as new workers to the existing cluster.

> Example: If the initial OCP deployment had three baremetal workers and the intended worker count was ten, current_worker_count would be 3 and scale_out_count would be 7. Scale out from three existing workers, adding seven new workers, for a total of ten worker nodes.

Reviewer 2: I think it would be helpful to add an example scale-out.yml:
**/root/jetlag/ansible/vars/scale_out.yml:**
```yaml
current_worker_count: 0
scale_out_count: 1
```

Author: Done as suggested

## Run mno-scale-out.yml

> Once the new worker records are added and Worker node variables are properly populated to the inventory and the scale_out.yml file has the proper values. The final step is to run the mno-scale-out.yml playbook.

Reviewer 2: ```suggestion
Once the new worker records are added and Worker node variables are properly populated in the inventory and the scale_out.yml file has the proper values. The final step is to run the mno-scale-out.yml playbook.
```

```console
(.ansible) [root@xxx-h01-000-r650 jetlag]# ansible-playbook -i ansible/inventory/cloud99.local ansible/mno-scale-out.yml
...
```

This playbook will:
- Generate node configuration yml

> - Invoke ```oc adm node-image create``` with the node configuration, which generates a discovery ISO

Reviewer 2: Consider using single backticks instead of triple backticks for inline commands
```suggestion
- Invoke `oc adm node-image create` with the node configuration, which generates a discovery ISO
```

- Boot the new worker nodes off of the generated discovery ISO
- Approve generated CSRs

> This workflow can be run repeatedly to add more workers to the existing cluster.

Reviewer 2: It would be nice to provide some estimate of how much time this step would take. In my case, it took more than an hour to scale 1 worker.

Author: Okk sure let me add as suggested

# Scale Out a Single-Node Openshift Deployment

> A JetLag deployed Single-Node Openshift deployment can be scaled out via JetLag. Workers can be added using JetLag Inventory and Playbooks. This guide assumes you have an existing Single-Node OCP cluster deployed via JetLag. The worker section in the JetLag inventory file should contain records that represent the worker nodes currently joined to the running cluster.

Reviewer 3: replace "currently joined" with "that will be joined"

## Addtional step in case of Multi SNO Environments

⚠️ A potential risk exists wherein a user who deploys a second Single Node OpenShift (SNO) instance could accidentally overwrite it if they later attempt to expand the first SNO with a worker node.

> ⚠️ In Multi SNO Environment, You can create a new json file with the bastion host and desired hosts in a order where Bastion node follows other desired hosts for each SNO Cluster. After creating a new json file, host this where your machine running the playbooks can reach and set the following var such that the modified ocpinventory json file is used, or specify a local path for the file:

Reviewer 3: Is this needed? We have it documented in the tips and vars document how to override an inventory file.
